### PR TITLE
Update UID only for WooCommerce cookies

### LINKED ISSUE #23682: nonce_user_logged_out filter breaks REST API cookie auth
**Describe the bug**
New filter for 'nonce_user_logged_out' (introduced in v3.6) breaks REST API cookie auth. REST API depends on `wp_create_nonce( 'wp_rest' );`  The nonce is created using the new session-based UID, but the rest_api_init (when it's verifying REST authentication) fails now because the session doesn't exist in the REST API.

**To Reproduce**
1. wp_create_nonce( 'wp_rest' ); inside another plugin
2. make an API call with the X-WP-Nonce value set to the value generated in step 1

**Expected behavior**
REST API call should authenticate successfully. Instead, 403 forbidden is returned because the "session" doesn't exist when WP receives the REST API request.

Solution is to modify the filter. Don't modify the UID if the $action is 'wp_rest'

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
not environment specific

### PR DESCRIPTION
**Pull request by claudiosanches**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

We need to filter into `nonce_user_logged_out` to prevent CSRF attacks for unregistered users, but we introduced some issues since changing this nonce will also invalidate all cookies because it relies in the customer UID.
This PR proposes to only change WooCommerce's cookies, so it doesn't affect 3rd party integrations, unless they are incorrectly using `woocommerce` as a prefix for any cookie or nonce.

Closes #23682.

### How to test the changes in this Pull Request:

I created a plugin to help test how nonces behave: https://gist.github.com/claudiosanches/71e4a1282599509720d72927d10a94ee
But you can also test and follow the same steps in #28364 to test with " WordPress Popular Posts".

1. Install and enable our test plugin: https://gist.github.com/claudiosanches/71e4a1282599509720d72927d10a94ee
2. Set the "Test nonce behavior" widget to any sidebar of your theme (just make sure you can access in WooCommerce's catalog page, you may have to change to a theme like Twenty Twenty-One or Storefront, I have tested with both just in case too).
3. Go to `/wp-admin/admin.php?page=wc-settings&tab=products` and "Enable AJAX add to cart buttons on archives".
4. Open a new incognito window in your browser and starts go your store's catalog page.
5. By clicking in `Test Request` it will display `true` in the bottom of the widget because a of successful request.
6. Now add any product to the cart, and note that is still working the `Test Request` button.
7. Refresh your catalog page and click in `Test Request` again, you'll see now an authentication error from the REST API, go to any page and check that `Test Request` returns the same error every time now.
8. Close your incognito window, apply this patch and start again from step 3.
9. Note that now our `Test Request` works as expected and always returns true, even after completing a checkout.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Prevent cart to reset all nonce in front-end.

### CODE REVIEW

File: includes/class-wc-session-handler.php [hunk 1]
```diff
@@ -187,6 +187,25 @@ public function generate_customer_id() {
 		return $customer_id;
 	}
 
+	/**
+	 * Get session unique ID for quests if session is initialized or user ID if logged in.
```
Comment (roykho) [inline includes/class-wc-session-handler.php:191]: Typo: `quests` should be `requests` ?

File: includes/class-wc-session-handler.php [hunk 2]
```diff
@@ -187,6 +187,25 @@ public function generate_customer_id() {
...
+	public function get_customer_unique_id() {
+		$customer_id = '';
+
+		if ( $this->has_session() && $this->_customer_id ) {
+			$customer_id = $this->_customer_id;
+		} elseif ( is_user_logged_in() ) {
+			$customer_id = (string) get_current_user_id();
+		}
+
+		return $customer_id;
```
Comment (roykho) [inline includes/class-wc-session-handler.php:206]: Return type is stated as string and only line 203 enforces that even though 201 may already be string. So perhaps remove `(string)` from line 203 and put it on line 206 `return (string) $customer_id` would be better?

File: includes/class-wc-session-handler.php [hunk 3]
```diff
@@ -288,13 +307,31 @@ public function forget_session() {
 	/**
 	 * When a user is logged out, ensure they have a unique nonce by using the customer/session ID.
 	 *
+	 * @deprecated 5.3.0
 	 * @param int $uid User ID.
-	 * @return string
+	 * @return int|string
 	 */
 	public function nonce_user_logged_out( $uid ) {
 		return $this->has_session() && $this->_customer_id ? $this->_customer_id : $uid;
```
Comment (roykho) [inline includes/class-wc-session-handler.php:315]: Maybe add `wc_doing_it_wrong()` to help developers change to new method in the future?

**Review by roykho — COMMENTED** [review]
Tested fine with the steps outlined in the PR. I also further tested Bookings Availability to make sure the REST API is working there.

Left some comments for improvement.

File: includes/class-wc-session-handler.php (hunk 2 above)
Comment (claudiosanches) [reply includes/class-wc-session-handler.php:206]: Only `get_current_user_id()` returns `int`, that's why I didn't want to cast string for everything by default.

Comment (claudiosanches) [reply includes/class-wc-session-handler.php:206]: See `generate_customer_id()`, there even if we use `get_current_user_id()` we still cast string.

File: includes/class-wc-session-handler.php (hunk 3 above)
Comment (claudiosanches) [reply includes/class-wc-session-handler.php:317]: I added a `wc_deprecated_function()`.

**Review by roykho — APPROVED** [review]
LGTM!

**Pull request merged**